### PR TITLE
Seed PageViews data with Remote API

### PR DESCRIPTION
**Pull request by kathleen-lund**
Creates the "remote" module in the project that can be run by first creating a .jar with the command `mvn clean package` while in the /remote directory, then running the .jar within the /target directory with the hostname (i.e. "<username>-new-step-2020.appspot.com", etc) as the final argument. Run the module after deploying the appengine project to seed the data. Must also now register all servlets in the `web.xml` file!

### CODE REVIEW

File: capstone/pom.xml [hunk 1]
```diff
@@ -108,6 +108,17 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>com.google.appengine</groupId>
+      <artifactId>appengine-remote-api</artifactId>
+      <version>1.9.64</version>
+    </dependency>
+
+    <dependency>
```
Comment (josh8551021) [inline capstone/pom.xml:117]: Should we perhaps replace the dependency on line 59 with this one just so that there's only one GAE SDK dependency?

File: capstone/src/main/java/com/google/step/servlets/RemoteServlet.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.google.step.servlets;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/remote-api")
```
Comment (josh8551021) [inline capstone/src/main/java/com/google/step/servlets/RemoteServlet.java:10]: Why is this needed? A little documentation can go a long way in this case.

File: capstone/src/main/java/com/google/step/servlets/RemoteServlet.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.google.step.servlets;
```
Comment (josh8551021) [inline capstone/src/main/java/com/google/step/servlets/RemoteServlet.java:1]: Nit: add licensing.

File: capstone/src/main/webapp/WEB-INF/web.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+<display-name>Restaurant Servlet</display-name>
+<servlet-name>RestaurantServlet</servlet-name>
+<servlet-class>com.google.step.servlets.RestaurantServlet</servlet-class>
+<load-on-startup>1</load-on-startup>
+</servlet>
+<servlet-mapping>
+<servlet-name>RestaurantServlet</servlet-name>
+<url-pattern>/restaurant</url-pattern>
+</servlet-mapping>
+</web-app>
```
Comment (josh8551021) [inline capstone/src/main/webapp/WEB-INF/web.xml:60]: Can all of this be tabified for readability purposes?

File: remote/pom.xml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                        <phase>package</phase>
+                        <goals>
+                            <goal>single</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <archive>
+                        <manifest>
+                            <mainClass>com.example.appengine.remote.RemoteApiExample</mainClass>
```
Comment (josh8551021) [inline remote/pom.xml:45]: RemoteApiExample? Should this be something else?

File: remote/src/main/java/com/example/appengine/remote/RemoteApiExample.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.GeoPt;
+import com.google.appengine.tools.remoteapi.RemoteApiInstaller;
+import com.google.appengine.tools.remoteapi.RemoteApiOptions;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class RemoteApiExample {
```
Comment (josh8551021) [inline remote/src/main/java/com/example/appengine/remote/RemoteApiExample.java:13]: This class should probably be renamed to something more descriptive.

File: remote/web/WEB-INF/web.xml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (josh8551021) [inline remote/web/WEB-INF/web.xml:1]: Is this web.xml required if there aren't any configuration changes specified?

File: capstone/pom.xml (hunk 1 above)
Comment (kathleen-lund) [reply capstone/pom.xml]: Yes, fixed!

File: capstone/src/main/java/com/google/step/servlets/RemoteServlet.java (hunk 2 above)
Comment (kathleen-lund) [reply capstone/src/main/java/com/google/step/servlets/RemoteServlet.java]: Haha, turns out it's not needed. It was in the example I was following so I thought it was necessary but it isn't! Deleted.

File: capstone/src/main/webapp/WEB-INF/web.xml (hunk 4 above)
Comment (kathleen-lund) [reply capstone/src/main/webapp/WEB-INF/web.xml]: Sorry about that, added correct formatting

File: remote/pom.xml (hunk 5 above)
Comment (kathleen-lund) [reply remote/pom.xml]: Yes, changed the title

File: remote/web/WEB-INF/web.xml (hunk 7 above)
Comment (kathleen-lund) [reply remote/web/WEB-INF/web.xml]: Not required, good catch! Deleted the file.

File: capstone/pom.xml [hunk 8]
```diff
@@ -40,7 +40,7 @@
     <dependency>
       <groupId>com.google.appengine</groupId>
       <artifactId>appengine-api-stubs</artifactId>
-      <version>1.9.59</version>
+      <version>1.9.64</version>
```
Comment (Bickor) [inline capstone/pom.xml:43]: I imagine we need to make this change to also use the remote api or is it just for being updated?

File: remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+      options = new RemoteApiOptions().server(serverString, 8080).useDevelopmentServerCredential();
+    } else {
+      options = new RemoteApiOptions().server(serverString, 443).useApplicationDefaultCredential();
+    }
+    RemoteApiInstaller installer = new RemoteApiInstaller();
+    installer.install(options);
+    try {
+      DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+      Entity pageViews1 = new Entity("PageViews");
+      pageViews1.setProperty("restaurantKey", 1);
```
Comment (Bickor) [inline remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: So in the end are we still going to use the restaurantKey? To my understanding we were changing to just use the key created by the entity itself.

Comment (kathleen-lund) [reply remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: I thought we were still putting restaurantKeys in the PageViews entities? They won't be in the RestaurantInfo entities anymore but do we need them for PageViews?

File: capstone/pom.xml (hunk 8 above)
Comment (kathleen-lund) [reply capstone/pom.xml:43]: I updated it to this because the simple remote API example I was following used this version so I figured it was probably best to update. Does that answer your question?

File: remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java (hunk 9 above)
Comment (kathleen-lund) [reply remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: Come to think of it, I don't think I need the keys for my visualization, so if we want to take that out I'm happy to. I think the MetricsClient would need to be updated but let me know what you think.

Comment (josh8551021) [reply remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: I'm not sure, but from my understanding, the only place that the entity key would be used is in the `RestaurantInfo` entity. Every other entity kind would just use the ID value extracted from the key of the `RestaurantInfo` entity.

Comment (josh8551021) [reply remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: That way, everything else would stay exactly as it was.

File: remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.GeoPt;
+import com.google.appengine.tools.remoteapi.RemoteApiInstaller;
+import com.google.appengine.tools.remoteapi.RemoteApiOptions;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class RemoteApiSeedPageViewData {
```
Comment (tzsebe) [inline remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:13]: Misc best practice: Declare a private default constructor, to prevent outsiders from constructing this, and to express that it's not mean to be used as an object you construct.

File: remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+      DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+      Entity pageViews1 = new Entity("PageViews");
+      pageViews1.setProperty("restaurantKey", 1);
+      pageViews1.setProperty("restaurantName", "Wildfire Restaurant");
+      pageViews1.setProperty("year", 2020);
+      pageViews1.setProperty("week", 26);
+      pageViews1.setProperty("count", 39);
+
+      Entity pageViews2 = new Entity("PageViews");
+      pageViews2.setProperty("restaurantKey", 1);
```
Comment (tzsebe) [inline remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:35]: Do these values correspond with the actual keys that are seeded for the restaurants themselves?

File: remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java (hunk 9 above)
Comment (mkmg) [reply remote/src/main/java/com/example/appengine/remote/RemoteApiSeedPageViewData.java:28]: Ideally there should only be 1 id/key used to uniquely identify the object. Since datastore generates a unique ID automatically, we should use that. However, I understand that we've already used the restaurant key string in many places, so if this would require a lot of work to change we can put in a todo and fix it later if there is time.
